pkg/placement/placement-resolver.go

```diff
@@ -0,0 +1,225 @@
+	// This function can fail if any object's GVR fails to be retrieved given
+	// its GVK, by the mapper shared with the placement-controller.
+	GeneratePlacementDecision(placementKey string) (*v1alpha1.PlacementDecisionSpec, error)
+	// ComparePlacementDecision compares the given placement decision spec
+	// with the maintained placement decision for the given placement key.
+	// The returned value is true iff:
+	//
+	// - The destinations in the PlacementDecisionSpec are an exact match
+	//of those in the resolution.
+	//
+	// - The same is true for every selected object.
+	//
+	// It is possible to output a false negative due to a temporary state of
+	// internal mappings being out of sync.
+	ComparePlacementDecision(placementKey string,
+		placementDecisionSpec *v1alpha1.PlacementDecisionSpec) bool
+	// NoteObject updates the maintained placement's objects resolution for the
+	// given placement key. If the object is being deleted, it is removed from
+	// the resolution if exists.
+	//
+	// If a key does not exist, it is created.
+	// The returned bool indicates whether the placement resolution was changed.
+	NoteObject(placementKey string, obj runtime.Object) (bool, error)
+	// RemoveObject removes the given object from the maintained placement's
+	// objects resolution for the given placement key.
+	//
+	// If a key does not exist in the cache, it is created.
+	// The returned bool indicates whether the placement resolution was changed.
+	RemoveObject(placementKey string, obj runtime.Object) bool
+	// SetDestinations updates the maintained placement's
+	// destinations data for the given placement key.
+	// If a key does not exist in the cache, it is created.
+	SetDestinations(placementKey string, destinations []string)
+	// DeleteResolution deletes the maintained placement resolution for
+	// the given placement key.
+	DeleteResolution(placementKey string)
+}
+
+func NewPlacementResolver(gvkGvrMapper util.GvkGvrMapper) PlacementResolver {
+	return &placementResolver{
+		gvkGvrMapper:          gvkGvrMapper,
+		placementToResolution: make(map[string]*placementResolution),
+	}
+}
+
+type placementResolver struct {
+	sync.RWMutex
+	gvkGvrMapper          util.GvkGvrMapper
+	placementToResolution map[string]*placementResolution
+}
+
+// GeneratePlacementDecision returns the placement decision for the given
+// placement key. If a key does not exist in the cache, it is created.
+// This function can fail if any object's GVR fails to be retrieved given
+// its GVK, by the mapper shared with the placement-controller.
+func (resolver *placementResolver) GeneratePlacementDecision(placementKey string) (
+	*v1alpha1.PlacementDecisionSpec, error) {
+	resolver.RLock() // lock for reading map
+	placementResolution, exists := resolver.placementToResolution[placementKey]
+	resolver.RUnlock() // unlock after accessing map
+
+	if !exists {
+		placementResolution = resolver.createResolution(placementKey)
+	}
```

Comment:
FYI: I see several repetitions of this sequence of statements, and also a variant that does not create if `!exists`. I usually DRY this out by replacing

```
placementResolution, exists := resolver.placementToResolution[placementKey]
```

with

```
placementResolution, _ := resolver.get(placementKey, true)
```

and replacing

```
	resolver.RLock() // lock for reading map
	placementResolution, exists := resolver.placementToResolution[placementKey]
	resolver.RUnlock() // unlock after accessing map
```

with

```
placementResolution, exists := resolver.get(placementKey, false)
```